From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: fjr <fjr@immortalwrt.org>
Date: Sat, 05 Apr 2026 12:00:00 +0800
Subject: [PATCH] pcie: mediatek-gen3: Add probe delay support

Add support for an optional "probe-delay-ms" device tree property that
delays PCIe port initialization. This is useful for endpoints like 5G/4G
modems that require extra time to complete power-up before the PCIe host
controller attempts link training.

The delay is inserted before the PHY and clock initialization, giving the
endpoint device sufficient time to stabilize after power-on.

Signed-off-by: fjr <fjr@immortalwrt.org>
---
 drivers/pci/controller/pcie-mediatek-gen3.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

--- a/drivers/pci/controller/pcie-mediatek-gen3.c
+++ b/drivers/pci/controller/pcie-mediatek-gen3.c
@@ -165,6 +165,7 @@ struct mtk_msi_set {
  * @num_clks: PCIe clocks count for this port
  * @wifi_reset: reset pin for WIFI chip
  * @wifi_reset_delay_ms: delay time for WIFI chip reset
+ * @probe_delay_ms: delay time before power-up for endpoint readiness
  * @irq: PCIe controller interrupt number
  * @saved_irq_state: IRQ enable state saved at suspend time
  * @irq_lock: lock protecting IRQ register access
@@ -189,6 +190,7 @@ struct mtk_gen3_pcie {
 	struct gpio_desc *wifi_reset;
 	u32 wifi_reset_delay_ms;
 
+	u32 probe_delay_ms;
 	int irq;
 	u32 saved_irq_state;
 	raw_spinlock_t irq_lock;
@@ -892,6 +894,9 @@ static int mtk_pcie_parse_port(struct mt
 		}
 	}
 
+	of_property_read_u32(dev->of_node, "probe-delay-ms",
+			     &pcie->probe_delay_ms);
+
 	return 0;
 }
 
@@ -1059,6 +1064,15 @@ static int mtk_pcie_setup(struct mtk_gen
 	reset_control_assert(pcie->mac_reset);
 	usleep_range(10, 20);
 
+	/* Delay before power-up if requested, to allow endpoint devices
+	 * (e.g. 5G/4G modems) to complete their own power-on sequence
+	 * before the host controller attempts link training.
+	 */
+	if (pcie->probe_delay_ms) {
+		dev_info(pcie->dev, "delaying probe by %u ms\n", pcie->probe_delay_ms);
+		msleep(pcie->probe_delay_ms);
+	}
+
 	/* Don't touch the hardware registers before power up */
 	err = pcie->soc->power_up(pcie);
 	if (err)
